src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,138 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ * @param isHostMemory - host or GPU memory
+ */
+case class UcxMemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie {
+  // Write this cookie to some address in memory
+  def writeToMemory(memory: UcxMemoryBlock)
+
+  // Size of this cookie in bytes
+  def size: Int
+}
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+/**
+ * Special case for metadata blocks, that always would be fetched using
+ * [[ UcxShuffleTransport.fetchBlocksByBlockIds() ]]. It has a method to get an actual blockId,
+ * it refers to,so when requested this MetadataBlockId, transport could know that soon blockId
+ * would be fetched, and may do some optimizations (pinning, pre-fetching, etc).
+ */
+trait MetadataBlockId extends BlockId {
+  def getBlockId: BlockId
+}
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: UcxMemoryBlock
+
+  // Called when the transport is done with the Memory, so we can unmap it for example
+  // the transport is not to use this Memory anymore
+  def doneWithMemory(mem: UcxMemoryBlock)
+}
+
+/**
+ * Async operation callbacks
+ */
+trait OperationCallback {
+  def onSuccess()
+  def onFailure(throwable: Throwable)
```

Comment:
We need in the callback:

1. Statistics of the request (`end`-`start` time of the request, bandwidth), and in the future info from UCX as to what path the request took NVLINK/IB/RoCE/TCP. This will enable metrics in the Spark UI to show the correct thing for the correct medium.

2. It seems to me that `Throwable` is the "we gave up here is something that happened" way to handle an error. Can we define a class that is something like `TransportError` that could help in debugging? I'd like to be able to match on the class and detect if it is a UCX error, a network error, a peer error/timeout, etc. It seems we could carry some of this with different instances of `Throwable` but I'd rather use it only in the case where we can't create a `TransportError` and some other error happened we don't have better handling for. What do you think?

Reply:
> Statistics of the request (end-start time of the request, bandwidth), and in the future info from UCX as to what path the request took NVLINK/IB/RoCE/TCP. This will enable metrics in the Spark UI to show the correct thing for the correct medium.

You can always add any statistics in the implementation class. Also since we have explicit progress for all operations `end-start` is not an informative metric. Since data may be received quickly and `progress` call was postponed, so you'll got callback later

Reply:
I think the point here is that UCX/Transport does not know how long a transfer was. Because it depends on the application to go and trigger `progress` which goes and finds out if the transfer is done in the hardware.

If that's the case, then yes there's nothing we can do. Ideally I'd like to know: how the transfer happened (was it NVLink, etc...), and I'd like to know how long it took, but maybe that second metric is not possible with UCX.